doc/sonic-application-extension/sonic-application-extension-guide.md

```diff
@@ -0,0 +1,216 @@
+
+rules/docker-XXX.mk
+```makefile
+$(DOCKER_XXX)_CONTAINER_PRIVILEGED = true
+$(DOCKER_XXX)_CONTAINER_VOLUMES += /etc/sonic:/etc/sonic:ro
+$(DOCKER_XXX)_CONTAINER_VOLUMES += /usr/share/sonic/scripts:/usr/share/sonic/scripts:ro
+$(DOCKER_XXX)_CONTAINER_TMPFS += /tmp/
+````
+
+These variables are used to generate manifest for docker at build time (see generate_manifest function in https://github.com/Azure/sonic-buildimage/blob/master/rules/functions):
+
+4. For extensions that provide CLI commands a CLI plugin is needed.
+
+4.1. Remove extension CLI commands from sonic-utilities code.
+
+4.2. Move the implementation as a separate file into sonic-buildimage under dockers/docker-xxx/cli folder.
+
+This plugin has to implement ```register``` function needed to be loaded by sonic-utilities core:
+
+dockers/docker-xxx/cli/show.py
+```python
+import click
+
+@click.command()
+def example()
+    pass
+
+def register(cli):
+    cli.add_command(example)
+```
+
+```register``` may throw an exception, in this case main CLI will still work but a warning will be printed to the user.
+
+
+4.3 UT for CLI:
+
+In case CLI has unit tests, they need to be moved from sonic-utilities into sonic-buildimage repository and placed under dockers/docker-xxx/cli-plugins-tests/.
+
+In case this folder exists tests are executed at docker image build time and test log is recorder in target/docker-xxx.gz.log.
+
+The command line to execute tests is:
+
+```
+pytest-3 -v
+```
+
+### Developing a new SONiC Application Extension
+
+To develop a new SONiC Application Extension use the following example extension as a template:
+https://github.com/stepanblyschak/sonic-example-extension.
+
+Prerequisites, build instructions and installation instructions are present in repository README.md file.
+
+### Adding 3rd party application to SONiC package database
+
+Modify files/build_templates/packages.json.j2 to include new package. Example:
+
+```json
+{
+    "cpu-report": {
+        "repository": "stepanblyschak/cpu-report",
+        "description": "CPU report example",
+        "default-reference": "1.0.0"
+    }
+}
+```
+
+### Building SONiC image with 3rd party application
+
+To build SONiC image with 3rd party application pre-installed use SONIC_PACKAGES target group.
+See https://github.com/Azure/sonic-buildimage/blob/master/rules/sonic-packages.mk.
+
+Create a file under rules/ called rules/cpu-report.mk with the following content:
+```makefile
+CPU_REPORT = cpu-report
+$(CPU_REPORT)_REPOSITORY = stepanblyschak/cpu-report
+$(CPU_REPORT)_VERSION = 1.0.0
+SONIC_PACKAGES += $(CPU_REPORT)
+```
+
+Additional options:
+
+```
+$(CPU_REPORT)_DEFAULT_FEATURE_STATE_ENABLED # "y" or "n" - whether feature is enabled by default at system start. Sets enabled in the FEATURE table. Disabled by default.
+$(CPU_REPORT)_DEFAULT_FEATURE_OWNER # "local" or "kube". Default is "local".
+```
+
+### Manifest Reference
+
+Label that manifest content should be written to:
+```
+com.azure.sonic.manifest
```

Comment:
@stepanblyschak writing this is google leads to many options. which one are you referring to?

Reply:
This is the label name. It is in the format of reverse domain notation which is following https://docs.docker.com/config/labels-custom-metadata/#key-format-recommendations,